Minor improvements around bpf map use and testing

lkpdn opened the pull request:
This PR prevents unexpected map regeneration triggered by user command and increases test coverage as well. Without this patch, the documentation for ConfigMap ``preallocate-bpf-maps`` option would have to be edited (see: https://github.com/cilium/cilium/blame/1d6f46b74074e40d61e8a09fcc3b09ad3445349a/Documentation/install/upgrade.rst#L338). Test coverage for map_linux increases 2% -> 61.3%.

Fixes: #7386 

```release-note
Fix possible map deletion via `cilium bpf policy { add | delete }`
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7538)
<!-- Reviewable:end -->

joestringer commented on pkg/bpf/map_linux.go:
```diff
@@ -869,22 +837,6 @@ func ConvertKeyValue(bKey []byte, bValue []byte, key interface{}, value interfac
 
-// MetadataDiff compares the metadata of the BPF maps and returns false if the
-// metadata does not match
-func (m *Map) MetadataDiff(other *Map) bool {
```
Good spotting, this is replaced by the objCheck() functionality inside of `OpenOrCreateMap()`.

lkpdn replied:
:)

joestringer commented on pkg/bpf/map_linux_test.go:
```diff
@@ -87,16 +113,270 @@ func (s *BPFPrivilegedTestSuite) TestOpen(c *C) {
+		DumpParserFunc).WithCache()
+	// not to mess up emitted test results, checking isNew is sufficient.
+	logLevel := logging.DefaultLogger.GetLevel()
+	logging.DefaultLogger.SetLevel(logrus.ErrorLevel)
```
This is really useful during test development, but do you think it's necessary in the tree?

If we want this everywhere, it can go into `runTests()`; otherwise we could remove them.

lkpdn replied:
Makes sense, will remove from the tree. Thanks.

lkpdn replied:
Ah no, it does not. I thought `DisableMapPreAllocation` could be out of the tree.
Setting log level to ErrorLevel during the whole test is acceptable?

joestringer replied:
Sure. We don't usually have explicit changing of log level in the tests, but error sounds fine to me.

borkmann replied:
@lkpdn Are you still planning on removing the setting of log level? Thx

lkpdn replied:
I was 50/50 about this but considering that nothing ensures that we'll never miss some important error message in these objCheck() call sites so removed it.

joestringer commented on pkg/bpf/map_linux_test.go:
```diff
@@ -87,16 +113,270 @@ func (s *BPFPrivilegedTestSuite) TestOpen(c *C) {
+		dump2[key.String()] = append(dump2[key.String()], "custom-"+value.String())
+	}
+	testMap.DumpWithCallback(customCb)
+	c.Check(dump2, DeepEquals, map[string][]string{
```
If you use `checker.DeepEquals` (from `pkg/datapath/checker`) for cases like this, then it will format errors nicely which makes debug a bit easier when there's any problem. Would you mind updating these?

lkpdn replied:
Ok I'll fix these. Thanks!

lkpdn replied:
Done (https://github.com/cilium/cilium/pull/7538/commits/8280912414b4b50800a879c2582e80a4fe08eb54)

joestringer commented on pkg/bpf/bpf.go:
```diff
@@ -101,6 +101,13 @@ func EnableMapPreAllocation() {
 
+// DisableMapPreAllocation disables BPF map pre-allocation as a default
+// setting. Some map types enforces pre-alloc strategy so this does not
+// take effect in that case.
```
Might be worth noting here as well that this only affects the creation of new maps, not existing maps.

lkpdn replied:
Ok, will push a fix later.

lkpdn replied:
Done (https://github.com/cilium/cilium/pull/7538/commits/8280912414b4b50800a879c2582e80a4fe08eb54)

joestringer commented on pkg/bpf/map_linux_test.go:
```diff
@@ -87,16 +113,270 @@ func (s *BPFPrivilegedTestSuite) TestOpen(c *C) {
+		0,
+		DumpParserFunc).WithCache()
+	isNew, err := existingMap.OpenOrCreate()
+	c.Check(err, IsNil) // Avoid assert to ensure Close() is called below.
```
For future reference, we often defer the `Close()` to handle cases like this. This is fine as-is, just letting you know in case you want to use that approach in future.

lkpdn replied:
IIUC all the `Check` except ones where testing `Close` itself is necessary could be changed to `Assert`, is that right? I didn't have much consideration for this so will fix all of these.

joestringer replied:
I think so, yeah. So long as cleanup is deferred it should be ok to use assert.

lkpdn replied:
In my last force-push I replaced all with `Assert` + `defer Close()` where testing `Close()` is unnecessary.

joestringer commented on pkg/bpf/map_linux_test.go:
```diff
@@ -87,16 +113,270 @@ func (s *BPFPrivilegedTestSuite) TestOpen(c *C) {
+	err = existingMap.Update(key1, value1)
+	c.Check(err, IsNil)
+	// key    val
+	// 100    200
```
Are these comments correct? `key` is defined above with the value 103, and the value is different too. Same for the similar comments below in the test.

lkpdn replied:
Simple mistake, thanks, I'll fix this.

lkpdn replied:
Done (https://github.com/cilium/cilium/pull/7538/commits/8280912414b4b50800a879c2582e80a4fe08eb54)

joestringer reviewed (changes requested):
This is awesome, thanks! I have a few comments below primarily over the testing code just for maintainability.

Commit:
Improve pkg/bpf/map_linux.go test coverage

Coverage increases 2.0% -> 61.3%.
Also includes minor cleanups of unused funcs and introduces
DisableMapPreAllocation() to allow toggling default fallback
setting for unrestrictive BPF map types which is useful
in test codebase.

Signed-off-by: Koichiro Den <den@klaipeden.com>

Commit:
Fix possible map deletion via `cilium bpf policy { add | delete }`

`cilium bpf policy` command could possibly trigger policy map
regeneration if the desired map attribute differs from the current
one. We better to just log the error if the map would not be present
for some reason. With this patch, the documentation for ConfigMap
``preallocate-bpf-maps`` option is now all correct about the map
regeneration.

Fixes: #7386 - `cilium bpf policy add` deletes maps if flags mismatch
Signed-off-by: Koichiro Den <den@klaipeden.com>
